### ISSUE TITLE
Corruption on ARMv8

### ISSUE THREAD

**Issue by shamisp** [open-mpi/ompi#2607]
Recently I have been testing OpenMPI on ARMv8 (I run on AMD
a1100) and I discovered that your commit
4b208e44637cd717aee6f0311873b4bfdfed4849

leads to unexpected low level failures on ARM:

"kernel: Unhandled fault: alignment fault (0x92000021) at
0x0000ffff95e29fff"

As an application I¹m running UCX perftest which uses the MPI to exchange
the data while all the rest communication happens in UCX.
(https://github.com/openucx/ucx/tree/master/src/tools/perf)

In order to build OpenMPI I used:
../configure --with-platform=optimized --prefix=/home/xxx/install/
--enable-orterun-prefix-by-default

And I use it as the following
mpirun -np 2 -npernode 2 -H node1 -x UCX_TLS=mm -x UCX_SHMEM_DEVICES=xpmem
/home/xxx/install/bin/ucx_perftest <options for atomic test>

If I go one commit before 4b208e44637cd717aee6f0311873b4bfdfed4849 (a.k.a
74a11d783296eff50fe1bab1a0a630500a63973c)
everything works without trouble.

I tried to look at 4b208e44637cd717aee6f0311873b4bfdfed4849 but this is a
pretty big change. I think this change corrupts some data such that it leads to alignment
failures (most like around atomics).

**Comment by ggouaillardet** [comment]
@shamisp which version of Open MPI are you using ?
i just found you need ```v2.x``` or ```master``` for ```aarch64``` support ...

**Comment by ggouaillardet** [comment]
@shamisp also, do you invoke ```mpirun``` on ```node1``` ?
if not, is ```mpirun``` invoked from an ```ARMv8``` node ?

regardless the issue, i am a bit surprised ```btl/tcp``` is even used.
both MPI tasks run on the same node, so an ```btl/sm``` or ```btl/vader``` are better fits.
they might disqualify themselves, (unless ```ucx``` blacklist them for a reason i ignore)

**Comment by shamisp** [comment]
@ggouaillardet
1. I use master.
2. I'm surprised either, but this is the first commit where the problem appears. UCX has not control over the runtime.

**Comment by rhc54** [comment]
I'm confused - the referenced changeset only touches the TCP/BTL. What does that have to do with UCX?

**Comment by hppritcha** [comment]
MPI is just used for out of and communication in these UCX unit tests.

**Comment by ggouaillardet** [comment]
@shamisp do you need special hardware to run UCX ?
if not, can you please provide basic instructions on how you built it ?
if some dependencies were needed (xpmem obviously), could you also detail these ?

**Comment by shamisp** [comment]
Update: if I convert the "new" structure to static (which is essentially similar to "global") the problem disappears. I suspect that with local variables we end up caching somewhere (still has to be identified) pointer to a stack, which leads to all kinda issues later on.

**Comment by ggouaillardet** [comment]
@shamisp i was unable to reproduce the issue on a raspberry pi3 running ```openSUSE Leap 42.2``` for ```aarch64```

note converting the structure to static is not a correct fix since the function must be thread safe (at least when the ```btl/tcp``` progress thread is used.
i reviewed the code and could not find any pointers left in the stack.
i also used ```valgrind``` to mark the structure as unaccessible, and could not hit any error

i noted several inconsistencies in your description
- ```btl/tcp``` should not even be used since you are running all MPI tasks on the same node
- your ```ucx``` configure command line contains ```--with-avx```, is this really used on ```ARMv8``` ?
- i am able to run ```ucx_perftest``` like this
```
mpirun -np 2 ./ucx_perftest -x mm -d posix -t am_lat
```
setting the ```UCX_TLS``` and ```UCX_SHMEM_DEVICES``` seems to have no effect (the app fails since it cannot open the device ""
the test name is also missing from your command line, and it seems mandatory
- your ```UCX_SHMEM_DEVICES``` suggests you are using ```xpmem```, but it is missing from your ```ucx``` command line. is it installed in ```/usr``` ? and in this case, i guess Open MPI is using ```xpmem``` too
- if you are really using ```xpmem```, where did you get it from and how did you configure it ?
- where is the error message coming from ? the app ? the kernel ? if the kernel, is it coming from ```xpmem```, ```knem``` or something else ?

bottom line, i am unable to reproduce the issue, and sadly, i cannot even make any sense from the description and the info you already provided.
i still hope i can understand what is going wrong and fix it shortly.

**Comment by jsquyres** [comment]
@ggouaillardet @shamisp This might be related to an issue that @opoplawski, our Fedora packager, is running in to.  Can you have a look at #2643?

**Comment by shamisp** [comment]
Please see inline

On Tue, Jan 3, 2017 at 7:50 AM, Gilles Gouaillardet <
notifications@github.com> wrote:

> @shamisp <https://github.com/shamisp> i was unable to reproduce the issue
> on a raspberry pi3 running openSUSE Leap 42.2 for aarch64
>

FYI, I'm working on a dedicated ARM regression server setup for OpenMPI and
other HPC related projects.
Hopefully in a near future it will be  much easier to get hands and a debug
setup.


> note converting the structure to static is not a correct fix since the
> function must be thread safe (at least when the btl/tcp progress thread
> is used.
> i reviewed the code and could not find any pointers left in the stack.
> i also used valgrind to mark the structure as unaccessible, and could not
> hit any error
>

Thanks for checking this.

> i noted several inconsistencies in your description
>
>    - btl/tcp should not even be used since you are running all MPI tasks
>    on the same node
>
>
Correct it is not user, but I'm sure it is initialized.

>
>    - your ucx configure command line contains --with-avx, is this really
>    used on ARMv8 ?
>
> Yes, this default "optimized" configure for UCX. On ARM and Power it is
ignored. (On ARM we use NEON and it is automatically detected)

>
>    - i am able to run ucx_perftest like this
>
> mpirun -np 2 ./ucx_perftest -x mm -d posix -t am_lat
>
>
Did you try to run any AMO tests ? I have observed the issue only with ucp
AMO.
I will provide the exact command shortly (for the moment I don't have an
access to the setup)


>
>    - your UCX_SHMEM_DEVICES suggests you are using xpmem, but it is
>    missing from your ucx command line. is it installed in /usr ? and in
>    this case, i guess Open MPI is using xpmem too
>
> It is autodetect for UCX  and I explicitly disabled in OMPI. The way OMPI
allocated XPMEM memory is x86 specific and does not work for other
platforms. This is one of the issues on my todo list.


>
>    - if you are really using xpmem, where did you get it from and how did
>    you configure it ?
>
> Yes, it was really used.
https://github.com/hjelmn/xpmem

>
>    - where is the error message coming from ? the app ? the kernel ? if
>    the kernel, is it coming from xpmem, knem or something else ?
>
> The error coming from kernel.


>
>
> bottom line, i am unable to reproduce the issue, and sadly, i cannot even
> make any sense from the description and the info you already provided.
>
I will do my best to provide better information.Shortly, I will post exact
command line for configure and execution.

**Comment by PHHargrove** [comment]
I have (non-root) access to systems with the following configurations:
```
8x2.4 GHz     aarch64 / 32 GB RAM / APM X-Gene Mustang board / Ubuntu 14.04.3 LTS
8x2 GHz       aarch64 / 16 GB RAM / AMD Opteron 1100 / openSUSE Leap 42.1
```
Both are bare hardware (not kvm or qemu, either of which might mask unaligned access issues).

If there are simple tests I can run to help track this problem, let me know.

**Comment by shamisp** [comment]
Okay - I actually narrowed down the issue to a bug in UCX (https://github.com/openucx/ucx/pull/1236). I have no idea why the problem show ups only after 4b208e4 but it is very consistent.
